msm/src/serialization/witness.rs

```diff
@@ -112,6 +140,43 @@ impl<Fp: PrimeField> Env<Fp> {
             }
         }
     }
+
+    pub fn add_rangecheck4_table_value(&mut self, i: usize) {
+        if i < (1 << 4) {
+            self.lookup_t_multiplicities_rangecheck4[i] += Fp::one();
+        } else {
+            self.lookup_t_multiplicities_rangecheck4[0] += Fp::one();
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.rangecheck15_lookups = vec![];
+        self.rangecheck4_lookups = vec![];
+    }
+
+    pub fn get_rangecheck4_normalized_multipliticies(
+        &self,
+        domain: EvaluationDomains<Fp>,
+    ) -> Vec<Fp> {
+        let mut m = vec![Fp::zero(); 1 << 4];
+        self.lookup_multiplicities_rangecheck4
+            .into_iter()
+            .zip(self.lookup_t_multiplicities_rangecheck4.iter())
+            .enumerate()
+            .for_each(|(i, (m_f, m_t))| m[i] = m_f / m_t);
+        let repeated_dummy_value: Vec<Fp> = iter::repeat(m[0])
+            .take((domain.d1.size - (1 << 4)) as usize)
+            .collect();
+        m.extend(repeated_dummy_value);
+        m
+    }
+
+    pub fn get_rangecheck15_normalized_multipliticies(
+        &self,
+        _domain: EvaluationDomains<Fp>,
+    ) -> Vec<Fp> {
+        self.lookup_multiplicities_rangecheck15.to_vec()
```

Comment:
Ah OK, answered my own question about why 15bit case does not have `t`. The multiplicities are /exactly/ equal to the "called" vector that we build in the witness, so they are much easier to compute.